src/storage/src/compaction/scheduler.rs

```diff
@@ -0,0 +1,449 @@
+            handler.run().await;
+        });
+        Self {
+            join_handle: Mutex::new(Some(join_handle)),
+            request_queue,
+            cancel_token,
+            task_notifier,
+        }
+    }
+
+    async fn remaining_requests(&self) -> usize {
+        self.request_queue.read().await.len()
+    }
+}
+
+#[allow(unused)]
+struct CompactionHandler<R, T: CompactionTask, P: Picker<R, T>> {
+    req_queue: Arc<RwLock<DedupDeque<TableId, R>>>,
+    cancel_token: CancellationToken,
+    task_notifier: Arc<Notify>,
+    limiter: Arc<CascadeRateLimiter<R>>,
+    picker: P,
+    _phantom_data: PhantomData<T>,
+}
+
+#[allow(unused)]
+impl<R, T: CompactionTask, P: Picker<R, T>> CompactionHandler<R, T, P> {
+    /// Runs table compaction requests dispatch loop.
+    pub async fn run(&self) {
+        let task_notifier = self.task_notifier.clone();
+        let limiter = self.limiter.clone();
+        loop {
+            tokio::select! {
+                _ = task_notifier.notified() => {
+                    // poll requests as many as possible until rate limited, and then wait for
+                    // notification (some task's finished).
+                    debug!("Notified, queue size: {:?}", self.req_queue.read().await.len());
+                    while let Some((table_id,  req)) = self.poll_task().await {
+                        if let Ok(token) = limiter.acquire_token(&req) {
+                            debug!("Executing compaction request: {}", table_id);
+                            self.handle_compaction_request(req, token).await;
+                        } else {
+                            // compaction rate limited, put back to req queue to wait for next
+                            // schedule
+                            debug!("Put back request {}, queue size: {}", table_id, self.req_queue.read().await.len());
+                            self.put_back_req(table_id, req).await;
+                            break;
+                        }
+                    }
+                }
+                _ = self.cancel_token.cancelled() => {
+                    info!("Compaction tasks scheduler stopped.");
+                    return;
+                }
+            }
+        }
+    }
+
+    #[inline]
+    async fn poll_task(&self) -> Option<(TableId, R)> {
+        let mut queue = self.req_queue.write().await;
+        queue.pop_front()
+    }
+
+    /// Puts request back to the front of request queue.
+    #[inline]
+    async fn put_back_req(&self, table_id: TableId, req: R) {
+        let mut queue = self.req_queue.write().await;
+        queue.push_front(table_id, req);
+    }
+
+    // Handles compaction request, submit task to bg runtime.
+    async fn handle_compaction_request(&self, mut req: R, token: RateLimitTokenPtr) -> Result<()> {
+        let cloned_notify = self.task_notifier.clone();
+        let task = self.build_compaction_task(req).await?;
+
+        common_runtime::spawn_bg(async move {
```

Comment:
We might also need a way to cancel these jobs. This will be useful while implementing `close_region()`, `drop_region()` and `stop()`

Reply:
Allow me add a todo here.